tests/test_validation.py

```diff
@@ -0,0 +1,237 @@
+    assert model.meta.object_attribute.string_attribute == "foo"
+
+    with pytest.warns(None) as warnings:
+        model.meta.object_attribute = None
+    assert len(warnings) == 0
+    assert model.meta.object_attribute.string_attribute is None
+
+
+def test_list_attribute_ssignment():
+    model = ValidationModel()
+
+    assert len(model.meta.list_attribute) == 0
+    with pytest.warns(None) as warnings:
+        model.meta.list_attribute.append({"string_attribute": "foo"})
+    assert len(warnings) == 0
+    assert model.meta.list_attribute[0].string_attribute == "foo"
+
+    with pytest.warns(ValidationWarning):
+        model.meta.list_attribute.append({"string_attribute": 42})
+    assert len(model.meta.list_attribute) == 1
+    assert model.meta.list_attribute[0].string_attribute == "foo"
+
+    with pytest.warns(None) as warnings:
+        model.meta.list_attribute = None
+    assert len(warnings) == 0
+    assert len(model.meta.list_attribute) == 0
+
+
+@pytest.mark.xfail(reason="validation of a required attribute not yet implemented", strict=True)
+def test_required_attribute_assignment():
+    model = RequiredModel()
+
+    with pytest.warns(None) as warnings:
+        model.meta.required_attribute = "foo"
+
+    with pytest.warns(ValidationWarning):
+        model.meta.required_attribute = None
+
+
+def test_fits_required_attribute_assignment():
+    model = FitsRequiredModel()
+
+    with pytest.warns(None) as warnings:
+        model.meta.required_keyword = "foo"
+    assert len(warnings) == 0
+    assert model.meta.required_keyword == "foo"
+
+    with pytest.warns(ValidationWarning, match="REQKWRD is a required value"):
+        model.meta.required_keyword = None
+    assert model.meta.required_keyword == "foo"
+
+    with pytest.warns(None) as warnings:
+        model.required_hdu = np.zeros((2, 2))
+    assert len(warnings) == 0
+    assert_array_equal(model.required_hdu, np.zeros((2, 2)))
+
+    with pytest.warns(ValidationWarning, match="1 is a required value"):
+        model.required_hdu = None
+    assert_array_equal(model.required_hdu, np.zeros((2, 2)))
+
+
+@pytest.mark.xfail(reason="validation of required attributes not yet implemented", strict=True)
+def test_validation_on_delete():
+    model = RequiredModel()
+
+    with pytest.warns(None) as warnings:
+        model.meta.required_keyword = "foo"
+    assert len(warnings) == 0
+
+    with pytest.warns(ValidationWarning):
+        del model.meta.required_keyword
+    assert model.meta.required_keyword == "foo"
+
+    model = RequiredModel(pass_invalid_values=True)
+
+    with pytest.warns(None) as warnings:
+        model.meta.required_keyword = "foo"
+    assert len(warnings) == 0
+
+    with pytest.warns(ValidationWarning):
+        del model.meta.required_keyword
+    assert model.meta.required_keyword is None
+
+
+def test_pass_invalid_values_attribute_assignment(monkeypatch):
+    def assert_passed(model, passed):
```

Comment:
This is essentially an assertion helper, so it might be good to set `__tracebackhide__ = True` in the function.

Reply:
I reworked this one to be parametrized as well, so the assertion helper disappeared.